src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs

```diff
@@ -1,107 +1,157 @@
-﻿using System.Diagnostics;
+﻿using System;
+using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
 using Orleans.GrainDirectory;
 using Orleans.SystemTargetInterfaces;
+using OutcomeState = Orleans.Runtime.GrainDirectory.GlobalSingleInstanceResponseOutcome.OutcomeState;
 
 namespace Orleans.Runtime.GrainDirectory 
 {
-    /// <summary>
-    /// A class that encapsulates response processing logic.
-    /// It is a promise that fires once it has enough responses to make a determination.
-    /// </summary>
-    internal class GlobalSingleInstanceResponseTracker : TaskCompletionSource<GlobalSingleInstanceResponseTracker.Outcome> {
-
-        public enum Outcome {
+    internal struct GlobalSingleInstanceResponseOutcome
+    {
+        public enum OutcomeState
+        {
             Succeed,
             RemoteOwner,
             RemoteOwnerLikely,
             Inconclusive
         }
 
+        public static readonly GlobalSingleInstanceResponseOutcome Succeed = new GlobalSingleInstanceResponseOutcome(OutcomeState.Succeed, default(AddressAndTag), null);
+
+        public readonly OutcomeState State;
+        public readonly AddressAndTag RemoteOwnerAddress;
+        public readonly string RemoteOwnerCluster;
+        public GlobalSingleInstanceResponseOutcome(OutcomeState state, AddressAndTag remoteOwnerAddress, string remoteOwnerCluster)
+        {
+            this.State = state;
+            this.RemoteOwnerAddress = remoteOwnerAddress;
+            this.RemoteOwnerCluster = remoteOwnerCluster;
+        }
+
+        public override string ToString()
+        {
+            return $"[{this.State} {this.RemoteOwnerAddress.Address}]";
```

Comment:
I think string interpolation inserts empty strings for null values, so this should be o.k.